core/templates/dev/head/components/CkEditorRteDirective.js

```diff
@@ -0,0 +1,219 @@
+                '</div></div>',
+      require: '?ngModel',
+
+      link: function(scope, el, attr, ngModel) {
+        var _RICH_TEXT_COMPONENTS = RteHelperService.getRichTextComponents();
+        var names = [];
+        var icons = [];
+        var canUseFs = ExplorationContextService.getPageContext() ===
+          PAGE_CONTEXT.EDITOR;
+        _RICH_TEXT_COMPONENTS.forEach(function(componentDefn) {
+          if (!((scope.uiConfig() &&
+            scope.uiConfig().hide_complex_extensions &&
+            componentDefn.isComplex) ||
+            (!canUseFs && componentDefn.requiresFs))) {
+            names.push(componentDefn.id);
+            icons.push(componentDefn.iconDataUrl);
+          }
+        });
+
+        /**
+         * Create rules to whitelist all the rich text components and
+         * their wrappers and overlays.
+         * See format of filtering rules here:
+         * http://docs.ckeditor.com/#!/guide/dev_allowed_content_rules
+         */
+        // Whitelist the component tags with any attributes and classes.
+        var componentRule = names.map(function(name) {
+          return 'oppia-noninteractive-' + name;
+        }).join(' ') + '(*)[*];';
+        // Whitelist the inline component wrapper, which is a
+        // span with a "type" attribute.
+        var inlineWrapperRule = ' span[type];';
+        // Whitelist the block component wrapper, which is a div
+        // with a "type" attribute and a CSS class.
+        var blockWrapperRule = ' div(oppia-rte-component-container)[type];';
+        // Whitelist the transparent block component overlay, which is
+        // a div with a CSS class.
+        var blockOverlayRule = ' div(oppia-rte-component-overlay);';
+        // Put all the rules together.
+        var extraAllowedContentRules = componentRule +
+                                       inlineWrapperRule +
+                                       blockWrapperRule +
+                                       blockOverlayRule;
+
+        var pluginNames = names.map(function(name) {
+          return 'oppia' + name;
+        }).join(',');
+        var buttonNames = [];
+        names.forEach(function(name) {
+          buttonNames.push('Oppia' + name);
+          buttonNames.push('-');
+        });
+        buttonNames.pop();
+        // All icons on the toolbar except the Rich Text components.
+        var allIcons = ['undo', 'redo', 'bold', 'Italic', 'numberedList',
+          'bulletedList', 'pre', 'indent', 'outdent'];
+
+        // Add external plugins.
+        CKEDITOR.plugins.addExternal(
+          'widget', '/third_party/static/ckeditor-widget-4.9.2/', 'plugin.js');
+        CKEDITOR.plugins.addExternal(
+          'lineutils',
+          '/third_party/static/ckeditor-lineutils-4.9.2/', 'plugin.js');
+        CKEDITOR.plugins.addExternal(
+          'sharedspace',
+          '/third_party/static/ckeditor-sharedspace-4.9.2/', 'plugin.js');
+        CKEDITOR.plugins.addExternal(
```

Comment:
I noticed you'd deleted the "pre" plugin from some places. Do you still need this? Ditto below under extraPlugins.

If so please add a comment to explain why the path here is different from the others. If not please remove pre from all places.

Reply:
Yeah, I need this. Done!